pkg/kstatus/polling/clusterreader/dynamic_reader.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2022 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package clusterreader
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DynamicClusterReader is an implementation of the ClusterReader that delegates
+// all calls directly to the underlying DynamicClient. No caching.
+type DynamicClusterReader struct {
+	DynamicClient dynamic.Interface
+	Mapper        meta.RESTMapper
+}
+
+func (n *DynamicClusterReader) Get(ctx context.Context, key client.ObjectKey, obj *unstructured.Unstructured) error {
+	mapping, err := n.Mapper.RESTMapping(obj.GroupVersionKind().GroupKind())
```

Comment:
FWIW, this DynamicClusterReader is only being used for GETing generated-resources, like ReplicaSets & Pods when a Deployment is being applied. So it's not being used for the primary objects the user cares about, and it's only being used for checking the objects status (tho it does get returned on the event, in case the caller wants to do something with it). Because it's not a primary object tho, we don't actually have the version, unless we assume generated resources with the same group also share the version (which doesn't help for Pods, which aren't in the apps group like ReplicaSets and Deployment).

But yes, if used more generically, I could see wanting to return the same version, if specified in the Unstructured obj.